### PR TITLE
Add block form event

### PR DESCRIPTION
**Pull request by Anatoliy057**

### CODE REVIEW

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java [hunk 1]
```diff
@@ -1519,4 +1519,78 @@ public boolean modifyEvent(String key, Mixed value, BindableEvent event) {
...
+					+ "{location: The location the block formed "
+					+ " | block: Type of old block state"
+					+ " | newblock: Type of new block state}"
+					+ "{} "
+					+ "{}";
+		}
+
+		@Override
+		public Driver driver() {
+			return Driver.EXTENSION;
```
Comment (PseudoKnight) [inline src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java:1546]: Wrong Driver.

**Comment by PseudoKnight** [comment]
Gave it some thought and I don't see a whole lot of value in getting the current block for the prefilter, or even the data for that matter. If you know the new block, you can infer the old block. And worst case scenario you can still get the old block with a function. Am I wrong?

**Comment by Anatoliy057** [comment]
For the prefilter of the old block, I agree, it is redundant since it can always be accurately determined by the result, but for the data the user will have to write the processing of all cases to determine the old block, and perhaps in the future there will be more of these cases.

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java (hunk 1 above)
Comment (Anatoliy057) [reply src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java]: Fixed

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java [hunk 2]
```diff
@@ -1519,4 +1519,75 @@ public boolean modifyEvent(String key, Mixed value, BindableEvent event) {
...
+		@Override
+		public String getName() {
+			return "block_form";
+		}
+
+		@Override
+		public String docs() {
+			return "{newblock: <string match> Type of new block state}"
+					+ "Called when a block is formed or spreads based on world conditions."
+					+ " Cancelling this event block will not be formed."
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java:1535]: `Cancelling this event will cause the block to not be formed.`

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java [hunk 3]
```diff
@@ -1519,4 +1519,75 @@ public boolean modifyEvent(String key, Mixed value, BindableEvent event) {
...
+					+ "{location: The location the block formed "
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java:1536]: `The location of the formed block`

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java [hunk 4]
```diff
@@ -1519,4 +1519,75 @@ public boolean modifyEvent(String key, Mixed value, BindableEvent event) {
...
+
+				return true;
+			}
+
+			return false;
+		}
+
+		@Override
+		public Map<String, Mixed> evaluate(BindableEvent e) throws EventException {
+			if(!(e instanceof MCBlockFadeEvent)) {
```
Comment (Pieter12345) [inline src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java:1575]: `MCBlockFadeEvent` -> `MCBlockFormEvent`
On this line, in the exception message and in the code below.

**Review by Pieter12345 — CHANGES_REQUESTED** [review]
Looking good overall. Using the wrong event for evaluation, which leads me to believe that you havn't tested it. Testing your code yourself is a great way to get bugs out early, and you should nearly always do that.

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java (hunk 2 above)
Comment (Anatoliy057) [reply src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java]: Fixed

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java (hunk 3 above)
Comment (Anatoliy057) [reply src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java]: Fixed

File: src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java (hunk 4 above)
Comment (Anatoliy057) [reply src/main/java/com/laytonsmith/core/events/drivers/BlockEvents.java]: Fixed

**Review by Pieter12345 — APPROVED** [review]

**Comment by PseudoKnight** [comment]
This looks good to me.

**Pull request merged**